### PR TITLE
Avoid concat in `array_append`

### PR DESCRIPTION
**Pull request by jayzhan211**
## Which issue does this PR close?

Follow up on https://github.com/apache/arrow-datafusion/pull/8108#pullrequestreview-1725755288

### CODE REVIEW

**Review by comphead — COMMENTED** [review]
Thanks @jayzhan211, looks good
Wondering if you can provide some numbers on how performance changed with the new method? 

You can create a small local benchmark on zillions of values to check before and after.

**Comment by jayzhan211** [comment]
I found the current change is 10x slower than the original one... I'm so surprise.

**Comment by jayzhan211** [comment]
unit: millis
| Number | This Change | Previous
| - | - | - |
| 2e8 | 391 | 252 |
| 5e8 | 1340 | 1533
| 1e9 | 16536 | 25460

**Comment by comphead** [comment]
Thanks @jayzhan211 
Numbers for two rows is definitely promising. 
However I'm not sure I'm following what is a single row or two row. 
perhaps you can attach a benchmark code to make it more clear?

**Comment by jayzhan211** [comment]
> Thanks @jayzhan211 Numbers for two rows is definitely promising. However I'm not sure I'm following what is a single row or two row. perhaps you can attach a benchmark code to make it more clear?

Single row is like [[1,2,3]] or ListArray( Int64Array(1,2,3) ).
Two row is like [[1,2,3], [4,5,6]] or ListArray (Int64Array(1,2,3), (4,5,6)) which is the target we are trying to optimized. We dont have to `compute::concat` each rows at the end instead we `extend` each row we need while iterating. This is why we can see `Tow row` in this change outperform the original code.

**Review by comphead — APPROVED** [review]
LGTM, thanks @jayzhan211

File: datafusion/physical-expr/src/array_expressions.rs [hunk 1]
```diff
@@ -579,57 +579,85 @@ pub fn array_pop_back(args: &[ArrayRef]) -> Result<ArrayRef> {
...
-            let new_values: Vec<_> = new_values.iter().map(|a| a.as_ref()).collect();
-            let values = arrow::compute::concat(&new_values)?;
-
-            Arc::new(ListArray::try_new(
-                Arc::new(Field::new("item", data_type.to_owned(), true)),
-                OffsetBuffer::new(offsets.into()),
-                values,
-                None,
-            )?)
+            return general_append_and_prepend(
```
Comment (alamb) [inline datafusion/physical-expr/src/array_expressions.rs:655]: giphy

**Review by alamb — APPROVED** [review]
Thank you @jayzhan211 and @comphead  -- this is so nice. It will be amazing if you could apply the same transformation to the rest of the functions in this module

**Pull request merged**